### PR TITLE
bpo-38328: Speed up the creation time of constant list and set literals.

### PR DESCRIPTION
**Pull request by brandtbucher**
At @methane's request, here is an alternate PR that takes place in the compiler, not the peephole optimizer. It would replace both #16498 and #16878.

I think I like this one better.

https://bugs.python.org/issue38328

### CODE REVIEW

**Comment by brandtbucher** [comment]
@benjaminp, care to take a look at this when you have a chance?

File: Lib/test/test_sys.py [hunk 1]
```diff
@@ -1222,7 +1222,7 @@ def get_gen(): yield 1
         # list
         samples = [[], [1,2,3], ['1', '2', '3']]
         for sample in samples:
-            check(sample, vsize('Pn') + len(sample)*self.P)
+            check(list(sample), vsize('Pn') + len(sample)*self.P)
```
Comment (methane) [inline Lib/test/test_sys.py:1225]: What is the purpose of this change?

Comment (brandtbucher) [reply Lib/test/test_sys.py:1225]: This is a `sys.getsizeof` test. The 3-element lists are overallocated, so we resize them here.

**Review by methane — APPROVED** [review]

**Comment by methane** [comment]
Would you merge master and `make regen-all`?

**Comment by brandtbucher** [comment]
Thanks @methane. This branch is no longer stale.

File: Python/compile.c [hunk 2]
```diff
@@ -3655,6 +3656,27 @@ starunpack_helper(struct compiler *c, asdl_seq *elts,
 {
     Py_ssize_t n = asdl_seq_LEN(elts);
     Py_ssize_t i, nsubitems = 0, nseen = 0;
+    if (n > 1 && are_all_items_const(elts, 0, n)) {
```
Comment (methane) [inline Python/compile.c]: Is `n > 1` really good threshold?  How about `n > 2`?

```
$ python3 -m timeit -s "a=frozenset((1,2))" -- "{*a}"
5000000 loops, best of 5: 63.4 nsec per loop
$ python3 -m timeit "{1,2}"
5000000 loops, best of 5: 61.5 nsec per loop

$ python3 -m timeit "[*(1,2)]"
5000000 loops, best of 5: 46.5 nsec per loop
$ python3 -m timeit "[1,2]"
5000000 loops, best of 5: 46.6 nsec per loop
```

Comment (brandtbucher) [reply Python/compile.c]: I only chose `n > 1` because it *always* emits fewer opcodes than before the change. If you think `n > 2` is a better heuristic, I'm fine with that.

Comment (brandtbucher) [reply Python/compile.c]: I think your first example is only slower because it pays the price of a `LOAD_NAME` instead of a `LOAD_CONST` (just a guess, though). It's not a direct comparison, unfortunately.

Comment (methane) [reply Python/compile.c]: The fewer opcodes is not always more efficient.
BUILD_LIST_UNPACK calls `PyList_New(0)` and `PyList_Extend(iterable)`.  It causes over-allocation.
On the other hand, BUILD_LIST just calls `PyList_New(n)`.

So I think we should be conservative about choosing the threshold.

Comment (brandtbucher) [reply Python/compile.c]: Thanks for the feedback @methane. I think this makes sense.

I've bumped the threshold to your recommended value of `n > 2`. Anything else?

**Comment by brandtbucher** [comment]
Thanks for the guidance and merge @methane!
